src/main/scala/fr/acinq/bitcoin/Psbt.scala

```diff
@@ -14,12 +14,186 @@ import scala.util.{Failure, Success, Try}
  *
  * @param global  global psbt data containing the transaction to be signed.
- * @param inputs  signing data for each input of the transaction to be signed.
- * @param outputs signing data for each output of the transaction to be signed.
+ * @param inputs  signing data for each input of the transaction to be signed (order matches the unsigned tx).
+ * @param outputs signing data for each output of the transaction to be signed (order matches the unsigned tx).
  */
 case class Psbt(global: Psbt.Global, inputs: Seq[Psbt.PartiallySignedInput], outputs: Seq[Psbt.PartiallySignedOutput]) {
+
+  import Psbt._
+
   require(global.tx.txIn.length == inputs.length, "there must be one partially signed input per input of the unsigned tx")
   require(global.tx.txOut.length == outputs.length, "there must be one partially signed output per output of the unsigned tx")
+
+  /**
+   * Implements the PSBT updater role; adds information about a given UTXO.
+   *
+   * @param inputTx         transaction containing the UTXO.
+   * @param outputIndex     index of the UTXO in the inputTx.
+   * @param redeemScript    redeem script if known and applicable.
+   * @param witnessScript   witness script if known and applicable.
+   * @param sighashType     sighash type if one should be specified.
+   * @param derivationPaths derivation paths for keys used by this UTXO.
+   * @return psbt with the matching input updated.
+   */
+  def update(inputTx: Transaction,
+             outputIndex: Int,
+             redeemScript: Option[Seq[ScriptElt]] = None,
+             witnessScript: Option[Seq[ScriptElt]] = None,
+             sighashType: Option[Int] = None,
+             derivationPaths: Map[PublicKey, KeyPathWithMaster] = Map.empty): Try[Psbt] = {
+    val outpoint = OutPoint(inputTx, outputIndex)
+    if (outputIndex >= inputTx.txOut.length) {
+      Failure(new IllegalArgumentException("output index must exist in the input tx"))
+    } else if (!global.tx.txIn.map(_.outPoint).contains(outpoint)) {
+      Failure(new IllegalArgumentException("psbt transaction does not spend the provided outpoint"))
+    } else {
+      val updatedInputs = global.tx.txIn.zip(inputs).map {
+        case (txIn, input) if txIn.outPoint == outpoint =>
+          val withUtxo = if (witnessScript.nonEmpty) {
+            input.copy(witnessUtxo = Some(inputTx.txOut(outputIndex)), redeemScript = redeemScript.orElse(input.redeemScript), witnessScript = witnessScript)
+          } else {
+            input.copy(nonWitnessUtxo = Some(inputTx), redeemScript = redeemScript.orElse(input.redeemScript))
+          }
+          withUtxo.copy(
+            sighashType = sighashType.orElse(input.sighashType),
+            derivationPaths = input.derivationPaths ++ derivationPaths
+          )
+        case (_, input) => input // we don't update other inputs
+      }
+      Success(this.copy(inputs = updatedInputs))
+    }
+  }
```

Comment:
I think that `update` would be more explicit if it was implemented with `Seq.indexWhere()` and `Seq.updated()`, and with a `Try` that around a simple method, for example:
```scala
def update(inputTx: Transaction,
             outputIndex: Int,
             redeemScript: Option[Seq[ScriptElt]] = None,
             witnessScript: Option[Seq[ScriptElt]] = None,
             sighashType: Option[Int] = None,
             derivationPaths: Map[PublicKey, KeyPathWithMaster] = Map.empty): Try[Psbt] = Try {

    require(outputIndex < inputTx.txOut.size, "output index must exist in the input tx")
    val outpoint = OutPoint(inputTx, outputIndex)
    val inputIndex = global.tx.txIn.indexWhere(_.outPoint == outpoint)
    require(inputIndex >= 0, "sbt transaction does not spend the provided outpoint")
    val input = this.inputs(inputIndex)
    val withUtxo = if (witnessScript.nonEmpty) {
      input.copy(witnessUtxo = Some(inputTx.txOut(outputIndex)), redeemScript = redeemScript.orElse(input.redeemScript), witnessScript = witnessScript)
    } else {
      input.copy(nonWitnessUtxo = Some(inputTx), redeemScript = redeemScript.orElse(input.redeemScript))
    }
    val withUtxo1 = withUtxo.copy(
      sighashType = sighashType.orElse(input.sighashType),
      derivationPaths = input.derivationPaths ++ derivationPaths
    )
    this.copy(inputs = this.inputs.updated(inputIndex, withUtxo1))
  }
```
WDYT ?

Same for `sign()`

Reply:
Good idea, done in https://github.com/ACINQ/bitcoin-lib/pull/43/commits/10066c9b5c63812bad463c0405583dbb07aa0bf7